Create a tutorial on how to build a Rhino application using shiny.fluent (#176)
Appsilon/shiny.fluent: vignettes/rhino-fluent.Rmd

---
title: "Tutorial: Build a Rhino application using shiny.fluent"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Tutorial: Build a Rhino application using shiny.fluent}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: sentence
---

## 1. Hello World with `rhino` and `shiny.fluent`

### Getting started with our development environment

Let us first install and create our `rhino` project structure.

To install `rhino`, please run the following on a R console in your working directory

```         
install.packages("rhino")
```

Now that `rhino` is installed, let us use it to initiate a project which will give us a structure

```         
rhino::init()

> ```

Reviewer: This is misleading. The most common case is that the user will use RStudio, thus they should use the Project  Wizard. There is no point in describing the process of creating a new project in detail, but you should mention all options and point to the documentation (main Rhino tutorial) where the user can find those details.

You will notice that the working directory now has a proper folder structure along with some other files.
Among those, you will also notice that `renv` is being used to manage the packages/dependencies for the project.
So, let us now install `shiny.fluent` into the project.

```         
renv::install("shiny.fleunt")
```

Using renv for installation benefits us.
In particular, this means that the local cache of package installations is used when possible.
This helps to avoid re-downloading packages that have already been downloaded before, and re-compiling packages from source when a binary copy of that package is already available.

Now, head over to the `dependencies.R` file and add `library(shiny.fluent)` to it.

> Upon saving the modified file, please take a snapshot of the dependencies so that the `renv.lock` file is updated.

Reviewer: Every change to a file needs to be provided as code snippet. See for example how this step was described in `shiny.i18n`: https://appsilon.github.io/shiny.i18n/articles/rhino.html#setup-rhino-application

```         
renv::snaphot()
```

Maintain this process whenever we need to install a new package in the project henceforth.

### Our first development on the app

-   Open the `app/main.R` file.

-   You will notice that currently it does not have any usage of `shiny.fluent` package to generate the text - "Hello".
    So let us introduce `shiny.fluent` and use it in our code to do the same action!

> ```         

Reviewer: Please add information about the edited file to the snippet (follow the style in e.g. Rhino main tutorial).

Author: fixed in latest commit

box::use(
  shiny[moduleServer, NS, renderText, tags, textOutput],
  shiny.fluent[fluentPage, Text]
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  fluentPage(

>     Text("Hello!", variant = "mega")

Author: This has been done in the later parts of the tutorial

>     Text("Hello!", variant = "mega")

Author: done in latest commit with new app structure

  )
}

#' @export
server <- function(id) {

>   moduleServer(id, function(input, output, session) {

Reviewer: Add a blank line below.

  }
  )
}
```

The above code uses [`fluentPage`](https://appsilon.github.io/shiny.fluent/reference/fluentPage.html) and [`Text`](https://appsilon.github.io/shiny.fluent/reference/Text.html) from `shiny.fluent` package to render the UI and text.
You will also notice the [`NS(id)`](https://shiny.rstudio.com/reference/shiny/latest/ns) used in the code to return a namespace function, which has been saved as `ns` and will be invoked later in the shiny module.

Now when you run the app, you should be seeing this -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_1.png")
```

Note: If you check the built-in testcases once changes are introduced, running `rhino::test_r()` will give you errors.

> To avoid failed github actions for now (in case you are pushing your code to Github), you may empty the test cases present in `tests\testthat\test-main.R`.

Reviewer: Please use the Unix path separator (/).

## 2. Show Data in a Table via a `rhino` module

So, in this section, we will introduce a table to our application that displays random imaginary sales deals.
The dataset comes from the `shiny.fluent` package itself - [fluentSalesDeals](https://appsilon.github.io/shiny.fluent/reference/fluentSalesDeals.html).

### Creating the module

-   In the `app\view\` folder, let us create a new file and name it `datatable.R`

-   This `app\view\datatable.R` file will be used to display the data table on the UI and we will call this module in the `app/main.R` file once done.
    Here is the code

>     ```         

Author: fixed in latest commit

    box::use(
      shiny[moduleServer, NS],
      shiny.fluent[fluentPage, Text, DetailsList]
    )

    #' @export
    ui <- function(id) {
      ns <- NS(id)

>         fluentPage(

Author: Agreed, just wanted to introduce the reader to `fluentPage` as part of the tutorial

>         fluentPage(

Author: introduced in latest commit

            Text("A randomly generated dataset of imaginary sales deals"),

>             DetailsList(items = shiny.fluent::fluentSalesDeals)

Author: resolved in latest commit

        )
    }

    #' @export
    server <- function(id) {

>       moduleServer(id, function(input, output, session) {

Reviewer: Add a blank line below.

      })
    }
    ```

So, here we are using [`DetailsList`](https://appsilon.github.io/shiny.fluent/reference/DetailsList.html) function to render the table.

### Calling the module

-   Go to `app/main.R` file and edit it with the following code

```         
box::use(
  shiny[moduleServer, NS],
  shiny.fluent[fluentPage, Text]
)

box::use(
  app/view/datatable,
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  fluentPage(
    Text("Hello!", variant = "mega"),
    datatable$ui(ns("datatable"))
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    datatable$server("datatable")
  }
  )
}
```

Notice here that we have called the `datatable` module by importing it to our main application file first.
It is now, that the main module will be able to use exported functions from `app/view/datatable.R`.

This code in the `app/main.R` file uses the concept of Shiny modules.
In the **`ui`** function, **`ns`** is used to create a unique namespace for the **`datatable`** element using the **`ns("datatable")`** syntax.

> This is important because when the UI contains multiple instances of the same element (e.g. multiple tables), each instance must have a unique namespace to prevent naming conflicts in the server function.

Reviewer: Instead of this section, just mention that Rhino applications use Shiny modules and refer to the documentation in Rhino and Shiny.

Author: * Added reference to documentations
* Keeping the `ns` explanation here, removed from other parts of the tutorial

In the **`server`** function, the **`ns`** function is not used directly, but its generated namespace is used as an argument for the **`datatable$server()`** function call.
This allows the server to reference the specific instance of the **`datatable`** element that is associated with the current session.

Now run the app.
You should be seeing something like this -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_2.png")
```

### Some improvements

#### Show informative columns with meaningful names

-   We will no more display all the columns that come with the dataset.
    Instead, we will display the `"rep_name", "date", "deal_amount", "city", "is_closed"` columns.

-   But for that we need to make the names more readable.
    So, we will assign proper names for every fields using the `tibble` package -

    ```         
    columns <- tibble(
        fieldName = c("rep_name", "date", "deal_amount", "city", "is_closed"),

>         name = c("Sales Rep", "Date", "Deal Amount", "City", "Is Closed?"))

Reviewer: Move the last parenthesis to the new line.

    ```

    Remember to include `tibble` inside `box::use()`, else the code will not be able to use the package.

-   Upon running the app with all these changes, you should see the improved table -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_3.png")
```

#### Render the Table from the server side inside the `datatable` module

-   Till now, we did not do anything inside the server side of the `app/view/datatable.R` file.
    Let us now use the server side to render the table rather than the UI side.
    This is particularly helpful when we want to introduce more features inside the table - such as user inputs.

-   So, here is the new modified `app/view/databale.R` file

    ```         
    box::use(
      shiny[moduleServer, NS, uiOutput, renderUI],
      shiny.fluent[fluentPage, Text, DetailsList],
      tibble[tibble]
    )

    columns <- tibble(
        fieldName = c("rep_name", "date", "deal_amount", "city", "is_closed"),
        name = c("Sales Rep", "Date", "Deal Amount", "City", "Is Closed?"))

    #' @export
    ui <- function(id) {
      ns <- NS(id)
        fluentPage(
            Text("A randomly generated dataset of imaginary sales deals"),

>             uiOutput(ns("datatable")))

Reviewer: Last parenthesis to the new line.

    }

    #' @export
    server <- function(id) {
      moduleServer(id, function(input, output, session) {
        output$datatable <- renderUI({
            DetailsList(items = shiny.fluent::fluentSalesDeals, columns = columns)
        })
      })
    }
    ```

The **`DetailsList`** table is being rendered inside the **`renderUI`** function, which generates UI output dynamically based on the **`items`** and **`columns`** arguments.
The **`renderUI`** function is triggered whenever there is a change in the input values or reactive expressions that depend on them (we will explore this further).

The **`columns`** data frame defines the field names and column names for the **`DetailsList`**.
The **`fieldName`** column contains the names of the columns in the **`shiny.fluent::fluentSalesDeals`** data frame, and the **`name`** column contains the display names for those columns.

The **`ns`** variable is used to generate namespace for the **`uiOutput`** function.

> This is important because it ensures that the input and output elements in the UI do not clash with elements from other modules or UI components in the same application.

Reviewer: This is a repetition, you already wrote about modules.

Author: removed this explanation as it is already discussed earlier

> Please notice that the `uiOutput` and `renderUI` functions are from the `shiny` library, not the `shiny.fluent` library.

Reviewer: Unnecessary

Author: removed

Also, the `app/main.R` remains unchanged, till now.

## 3. Add a simple toggle input feature for the table

Till now, we could only view table with all the data it has.
We will now add a functionality to view those rows in the table where value of the "Is Closed?" column is 1 by default and provide a toggle to the user to also include results where the value is 0.
The "Is Closed?" column basically indicates whether the deal is closed or not with values 0 and 1 where 0 means False/No and 1 means True/Yes.

> -   Let us first install the `dplyr` package.

Reviewer: Please install all required packages in a single step at the beginning of the tutorial.

Author: The tutorial unwinds with a theme of demand and supply, hence introduced new packages only when they were required.

Suggesting mentioning this point to install all packages at the beginning of the tutorial at the start/end of the tutorial

Reviewer: If you want to add each package separately, provide complete information - remind how to install, provide updated `dependencies.R`, and finally snapshot call.

    Please follow the same process we followed for `shiny.fluent` installation.

-   We will use [`Toggle.shinyInput`](https://appsilon.github.io/shiny.fluent/reference/Toggle.html) from `shiny.fluent` package which represents a physical switch that allows someone to choose between two mutually exclusive options.For example,"On/Off","Show/Hide".Choosing an option should produce an immediate result.

-   In the UI, we include the code for showing the toggle for input - `Toggle.shinyInput(ns("includeOpen"), label = "Include open deals"),` and work on the input provided inside the server section

> ```         

Reviewer: Provide a snippet with the whole file content.

#' @export
ui <- function(id) {
  ns <- NS(id)
    fluentPage(
        Text("A randomly generated dataset of imaginary sales deals"),
        Toggle.shinyInput(ns("includeOpen"), label = "Include open deals"),
        uiOutput(ns("datatable")))
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    
    filtered_deals <- reactive({
      shiny.fluent::fluentSalesDeals |> filter(
        is_closed | input$includeOpen
      )
    })

    output$datatable <- renderUI({
        DetailsList(items = filtered_deals(), columns = columns)
    })
  })
}
```

This is how the app looks now!

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_gif_1.gif")
```

## 4. Improve visuals for the table with elevated cards

We can create cards with elevations just like in Fluent UI [documentation](https://developer.microsoft.com/en-us/fluentui#/styles/web/elevation).
Let us style the data table area.

The following code does 2 things -

-   Creates an elevated card around the data table area

-   Confines the height of the table area

```         
div(
      class = "ms-depth-64", 

>       style = "height: 500px; overflow: auto;",

Reviewer: Inline styling is a bad practice! For styling, use Sass setup provided by Rhino.

Author: added comment to redirect user to custom styling in rhino documentation

      uiOutput(ns("datatable"))

>     )

Reviewer: 1. Provide the whole file snippet.
2. Fix indentation.

```

The above code creates a **`div`** element with the following attributes -

**`class = "ms-depth-64"`**assigns the CSS class **`ms-depth-64`** to the **`div`** element, which may define the element's visual appearance and behavior.

**`style = "height: 500px; overflow: auto;"`** sets the inline CSS style for the **`div`** element.
Specifically, it sets the height of the **`div`** element to be 500 pixels and enables scrolling within the **`div`** element when its contents overflow its boundaries.

So, this is how the UI should look now -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_gif_2.gif")
```

## 5. Render a Bar Chart with logic in a helper function

We will now add a bar chart that plots the deal amounts closed by the sales reps. For this, we will code a helper function and keep it inside the `app/logic` folder.

### Helper function in the logic folder of the `rhino` project

-   Go to the `app/logic` folder and create a new `.R` file.
    Lets name it `barchart.R`

> -   Install `gglplot2` and `plotly` packages for this purpose.

Author: The tutorial unwinds with a theme of demand and supply, hence introduced new packages only when they were required.

Suggesting mentioning this point to install all packages at the beginning of the tutorial at the start of the tutorial

    )

    #' @export
    generate_barchart <- function(filtered_data) {
            ggplot(filtered_data, aes(x = rep_name, y = deal_amount, fill = rep_name)) +
            geom_bar(stat = "identity") +
            labs(x = "Sales Rep Name", y = "Deal Amount") +
            theme_light()
    }
    ```

-   The above code block loads the **`ggplot2`** package and imports several functions from it.
    It then defines a function called **`generate_barchart()`** that takes a **`filtered_data`** argument as input.
    The function uses the **`ggplot()`** function to create a bar chart with **`x = rep_name`** and **`y = deal_amount`**, where **`rep_name`** and **`deal_amount`** are columns in the **`filtered_data`** data frame.
    The **`fill`** aesthetic is also set to **`rep_name`**, which maps each bar to a unique color based on the **`rep_name`** variable.
    The function also adds labels to the x and y axes and applies the **`theme_light()`** theme to the plot.
    Finally, the function is exported with **`#' @export`**, which makes it available to other packages and scripts that import this one.

### Call the helper function in the desired `rhino` module

> -   Now let us go back to the `app\view\datatable.R` file.

Reviewer: Why the chart is not in a separate module?

Author: The logic has been separated as a standalone function inside logic folder. Did not create a separate module for this in view folder to -
- reduce complexity
- include an inclusive example to call a logic function in a view module that gets called in main file

    Here we need to call the `generate_barchart()` function we just created along with the required dependencies to render it on UI as a plotly bar chart.

-   So, the code begins with

    ```         
    box::use(
      shiny[moduleServer, NS, uiOutput, renderUI, reactive, div],
      shiny.fluent[fluentPage, Text, DetailsList, Toggle.shinyInput],
      tibble[tibble],
      dplyr[filter],
      plotly[plotlyOutput, renderPlotly],
    )

    box::use(
      app/logic/barchart[generate_barchart],
    )
    ```

    Here, we have called the `plotly` library, and also the function that we defined in the `app/logic` folder.

-   Now, we will add one more `div` in the ui section after the datatable to make room for the bar chart

> ````         

Author: Can we make an exception here? I have explained the code in chunks and provided the whole code at the end

```         
 div(
      class = "ms-depth-64",

>       style = "height: 500px; overflow: auto;",

Reviewer: Again, avoid inline styling.

Author: added comment to redirect user to custom styling in rhino documentation

      plotlyOutput(ns("barchart"))
      )
```

`plotlyOutput` will help us to render the graph
````

-   Similarly, in the server section, we will need to add some code to ask the backend of the application for rendering the barchart

    ```         
     output$barchart <- renderPlotly({
            generate_barchart(filtered_deals())
        })
    ```

Once ready, on running the application, we should see the plotly bar chart rendering successfully and also responding to our interactions with the toggle input button.

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_4.png")
```

## 6. Some Improvements for better User Experience

> ### Keep the data table and bar chart on the same row

Reviewer: Why you are not using Fluent components like Card as containers for the application elements?

Author: with reference to the this [tutorial](https://youtu.be/KKWAeeZM4Rc?t=1793), card looks like a function created to build containers. Hence directly demonstrated with divs inside code.

Reviewer: And adding a custom component based on Fluent is a perfect fit for the tutorial, so you can reuse it.

This can be done by wrapping the `div`s into a parent `div` and using the **`container-fluid`** class to ensure that the **`div`**s fill the screen horizontally.

```         
div(
          class = "container-fluid",
          style = "margin-left: 0px; margin-right: 0px;",
          div(
              class = "ms-depth-8",
              style = "height: 500px; overflow: auto; width: 50%; display: inline-block;",
              uiOutput(ns("datatable"))
          ),
          div(
              class = "ms-depth-8",
              style = "height: 500px; overflow: auto; width: 50%; display: inline-block;",
              plotlyOutput(ns("barchart"))
          ),
          style = "display: flex; flex-wrap: wrap;"
        )

> ```

Reviewer: Avoid inline styling.

Author: added comment to redirect user to custom styling in rhino documentation

We also set the **`margin-left`** and **`margin-right`** to 0 pixels to remove any horizontal margin that might exist.
To make the **`div`**s display side by side, we set the **`display`** property to **`inline-block`**.
We also set the width of each **`div`** to **`50%`** to ensure that they fill half the screen each.

### Change orientation of X axis labels in plotly bar chart

```         
#' @export
generate_barchart <- function(filtered_data) {
        ggplot(filtered_data, aes(x = rep_name, y = deal_amount, fill = rep_name)) +
        geom_bar(stat = "identity") +
        labs(x = "Sales Rep Name", y = "Deal Amount") +

>         theme_light()+

Reviewer: missing whitespace

```         
# app/main.R

box::use(
  shiny[moduleServer, NS, div],
  shiny.fluent[fluentPage, Text]
)

box::use(
  app/view/datatable
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  fluentPage(      
    Text("Sales Data Dashboard", variant = "xxLarge"),
    datatable$ui(ns("datatable"))
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    datatable$server("datatable")
  })
}
```

This code in the `app/main.R` file is based the concept of [Shiny modules](https://shiny.rstudio.com/articles/modules.html) where [box modules in rhino](https://appsilon.github.io/rhino/articles/explanation/box-modules.html) are used.

Notice here that we have called the `datatable` module by importing it to our main application file first.
It is now, that the `main` module will be able to use exported functions from `app/view/datatable.R` module.

Now run the app.
You should be seeing something like this -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_2.png")
```

### Some improvements

> #### Show informative columns with meaningful names

Reviewer: Having a section called in a meaningless way: "Some improvements" makes no sense, especially since it contains only one subsection.

  name = c("Sales Rep", "Date", "Deal Amount", "City", "Is Closed?")
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  div(
    Text("A randomly generated dataset of imaginary sales deals", variant = "large"),
    div(
      class = "ms-depth-4",
      Toggle.shinyInput(ns("includeOpen"), label = "Include open deals")
    ),
    uiOutput(ns("datatable")),
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    filtered_deals <- reactive({
      shiny.fluent::fluentSalesDeals |> filter(
        is_closed | input$includeOpen
      )
    })
    output$datatable <- renderUI({
      DetailsList(items = filtered_deals(), columns = columns)
    })
  })
}
```

This code is defining a reactive expression called `filtered_deals` in an R Shiny application.

The reactive expression uses the `|>` operator to pipe the `shiny.fluent::fluentSalesDeals` data frame to the `filter()` function which comes from the `dplyr` package.
The `filter()` function is used to filter rows from a data frame based on one or more conditions.

In this case, the `filter()` function is filtering rows from `shiny.fluent::fluentSalesDeals` based on two conditions:

> -   `is_closed`: This is a logical column in the `shiny.fluent::fluentSalesDeals` data frame that indicates whether a sales deal has been closed or not.

Reviewer: This part is too repetitive, I don't see a point in elaborating that function named `filter` does filtering.

-   `input$includeOpen`: This is a reactive input object that is defined in the UI part of the code.
    It is used to determine whether open deals (i.e. deals that have not been closed) should be included in the filtered data or not.

The `|` symbol between the two conditions is the logical OR operator, which means that the `filter()` function will return all rows where either `is_closed` is TRUE or `input$includeOpen` is TRUE.

> Also, please notice that we have added the style for `class = "ms-depth-4"` in the `app/style/main.scss` file. The [`ms-depth-8`](https://developer.microsoft.com/en-us/fluentui#/styles/web/elevation) class is a predefined core class from Microsoft's Fluent UI which provides elevation to the div, `ms-depth-4` being the lowest.

Reviewer: ```suggestion
Also, please notice that we have added an additional style to the class `ms-depth-4` in the `app/style/main.scss` file. The [`ms-depth-8`](https://developer.microsoft.com/en-us/fluentui#/styles/web/elevation) class is a predefined core class from Microsoft's Fluent UI which provides elevation to the div, `ms-depth-4` being the lowest.
```

Here is the `app/style/main.scss` file now -

```

> # app/style/main.scss

Reviewer: This is not a valid comment in `scss` file!

> .ms-depth-4{

Reviewer: ```suggestion
.ms-depth-4 {
```

      span(class = "ms-depth-8", uiOutput(ns("datatable")))
    )
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    filtered_deals <- reactive({
      shiny.fluent::fluentSalesDeals |> filter(
        is_closed | input$includeOpen
      )
    })
    output$datatable <- renderUI({
      DetailsList(items = filtered_deals(), columns = columns)
    })
  })
}
```
The `Stack` element in the code allows for stacking multiple elements vertically or horizontally.

The horizontal argument is set to TRUE to horizontally stack the two elements. The tokens argument is a list that specifies the padding and the gap between the two elements. In this case, the padding is set to 20 and the childrenGap is set to 10. This will add a padding of 20 pixels and a gap of 10 pixels between the other `span` of the `Stack` element.

The span element has a class attribute set to `ms-depth-8`. This is a class provided by the Microsoft fluent UI that adds a shadow effect to the element, giving it a depth effect. The `uiOutput` element is used to render the `DetailsList` element that contains the filtered data.

> The style for the class `ms-depth-8` is added in the `app/style/main.scss` file. So the file now is modified to -

Reviewer: ```suggestion
The additional style for the class `ms-depth-8` is added in the `app/style/main.scss` file. So it now is modified to -
```

```

> # app/style/main.scss

Reviewer: Not a valid comment.

> .ms-depth-4{
>     display: flex; 
>     flex-wrap: wrap;
>     padding: 5px;
> }
> 
> .ms-depth-8{
>     height: 500px; 
>     overflow: auto; 
>     width: 50%; 
>     display: inline-block;
> }

Reviewer: ```suggestion
.ms-depth-4 {
    display: flex; 
    flex-wrap: wrap;
    padding: 5px;
}

.ms-depth-8 {
    height: 500px; 
    overflow: auto; 
    width: 50%; 
    display: inline-block;
}
```

* We shall use the `gglplot2` and `plotly` packages (which we installed in the beginning) for this purpose. 

Here is the code -

```
# app/logic/barchart.R

box::use(
  ggplot2[ggplot, aes, geom_bar, labs, theme_light, theme, element_text],
)

#' @export
generate_barchart <- function(filtered_data) {
        ggplot(filtered_data, aes(x = rep_name, y = deal_amount, fill = rep_name)) +
        geom_bar(stat = "identity") +
        labs(x = "Sales Rep Name", y = "Deal Amount") +
        theme_light() +
        theme(axis.text.x = element_text(angle = 60, hjust = 1))
}
```

The above code block loads the `ggplot2` package and imports several functions from it. It then defines a function called `generate_barchart()` that takes a `filtered_data` argument as input. The function uses the `ggplot()` function to create a bar chart with `x = rep_name` and `y = deal_amount`, where `rep_name` and `deal_amount` are columns in the `filtered_data` data frame. The `fill` aesthetic is also set to `rep_name`, which maps each bar to a unique color based on the `rep_name` variable. The function also adds labels to the x and y axes and applies the `theme_light()` theme to the plot.

> The `element_text()` function takes several arguments to customize the text appearance, including the `angle` argument to specify the rotation angle in degrees. Here, we set the angle to 60 degrees to rotate the labels slightly vertically, and we set the `hjust` argument to 1 to align the labels to the right of the tick marks.

Reviewer: A general description should be enough, this is not a tutorial for `ggplot2`.

> ### Call the helper function in the desired `rhino` module

Reviewer: The chart should have its own module. And this is the opportunity to also move the toggle to a separate one. Check the structure in the main Rhino tutorial.

In the UI, we include the code for showing the toggle for input - `Toggle.shinyInput(ns("includeOpen"), label = "Include open deals"),` and work on the input provided inside the server section

So, the `app/view/datatable.R` file now is modified to -

```         
# app/view/datatable.R

box::use(
  dplyr[filter],
  shiny[div, moduleServer, NS, reactive, renderUI, uiOutput],
  shiny.fluent[DetailsList, Text, Toggle.shinyInput],
  tibble[tibble],
)

columns <- tibble(
  fieldName = c("rep_name", "date", "deal_amount", "city", "is_closed"),
  name = c("Sales Rep", "Date", "Deal Amount", "City", "Is Closed?")
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  div(
    Text("A randomly generated dataset of imaginary sales deals", variant = "large"),
    div(
      class = "ms-depth-4",
      Toggle.shinyInput(ns("includeOpen"), label = "Include open deals")
    ),
    uiOutput(ns("datatable")),
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    filtered_deals <- reactive({

>       shiny.fluent::fluentSalesDeals |> filter(

Reviewer: The comment about the usage of `::` is lost.

        is_closed | input$includeOpen
      )
    })
    output$datatable <- renderUI({
      DetailsList(items = filtered_deals(), columns = columns)
    })
  })
}
```
The `Stack` element in the code allows for stacking multiple elements vertically or horizontally.

The horizontal argument is set to TRUE to horizontally stack the two elements. The tokens argument is a list that specifies the padding and the gap between the two elements. In this case, the padding is set to 20 and the childrenGap is set to 10. This will add a padding of 20 pixels and a gap of 10 pixels between the other `span` of the `Stack` element.

The span element has a class attribute set to `ms-depth-8`. This is a class provided by the Microsoft fluent UI that adds a shadow effect to the element, giving it a depth effect. The `uiOutput` element is used to render the `DetailsList` element that contains the filtered data.

The additional style for the class `ms-depth-8` is added in the `app/style/main.scss` file. So it now is modified to -
```
// app/style/main.scss

.ms-depth-4 {
    display: flex; 
    flex-wrap: wrap;
    padding: 5px;
}

.ms-depth-8 {
    height: 500px; 

>     overflow: auto; 

Reviewer: This results in the resize of the table that happens after everything is rendered. Changing to `overflow: overlay;` should help. Unfortunately, you will need also to regenerate gifs.

    display: inline-block;
}
```

The Sass code limits the height of the table to 500 pixels, width to 50% of the screen size and allows scrolling through the data. Similar to the previous usecase, we would again have to repeat the Sass building process with `rhino` as the file has been updated.

```
# in R console
rhino::build_sass()
```

So, this is how the UI should look now -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_gif_2.gif")
```
Note that we now we have space beside the table for additional UI component(s).

## Create separate `rhino` modules for toggle input and datatable

Now that we have successfully introduced a datatable and a toggle input in our application, let us further create different Rhino modules for them!

We already have the datatable module with us. We will thus create one more module - `app/view/toggle.R` and make some modifications in `app/view/datatable`. 

These modules will be called from the main file.

### The `app/view/toggle.R` module -

```
# app/view/toggle.R

box::use(
  dplyr[filter],
  shiny[div, moduleServer, NS, reactive],
  shiny.fluent[Text, Toggle.shinyInput],
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  div(

>     Text("A randomly generated dataset of imaginary sales deals", variant = "large"),

Reviewer: Why the subtitle is a part of the toggle module?

  moduleServer(id, function(input, output, session) {
    output$datatable <- renderUI({
      DetailsList(items = data(), columns = columns)
    })
  })
}
```
Please notice, that inside the server function, the response obtained from the `toggle` module earlier is also passed on to be used in displaying the data in a table. 

We are not returning any information from this module to the main module as it's sole purpose is to display given data in a table with desired columns in an easily readable form.

### The modified `app/main.R` module

Since we modified the modules, `app/main.R` also has to be modified.

```
# app/main.R

box::use(
  shiny[moduleServer, NS, observe, span],
  shiny.fluent[fluentPage, Stack, Text],
)

box::use(
  app/view/datatable,
  app/view/toggle,
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  fluentPage(      
    Text("Sales Data Dashboard", variant = "xxLarge"),
    toggle$ui(ns("toggle")),
    Stack(
      horizontal = TRUE,
      tokens = list(padding = 20, childrenGap = 10),
      span(class = "ms-depth-8", datatable$ui(ns("datatable")))
    )
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {

>     observe({

Reviewer: The `observe` is redundant.

      data <-toggle$server("toggle")
      datatable$server("datatable", data)
    })
  })
}
```
Here we have used the `toggle` and `datatable` modules and called them from the code.

## Create a `rhino` module to render a bar chart with logic in a helper function

We will now add a bar chart that plots the summarized amount of deals closed by the sales reps. For this, we will code a helper function and keep it inside the `app/logic` folder.

### Define the helper function

Go to the `app/logic` folder and create a new `.R` file. Lets name it `barchart.R`. We shall use the `gglplot2` and `plotly` packages now (which we installed in the beginning) for this purpose. 

Here is the code -

```
# app/logic/barchart.R

box::use(
  ggplot2[aes, element_text, geom_bar, ggplot, labs, theme, theme_light],
)

#' @export
generate_barchart <- function(filtered_data) {
  ggplot(filtered_data, aes(x = rep_name, y = deal_amount, fill = rep_name)) +
    geom_bar(stat = "identity") +
    labs(x = "Sales Rep Name", y = "Deal Amount") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 60, hjust = 1))
}
```

> The above code block loads the `ggplot2` package and generates a barchart with names of the sales reps on the X axis and the  summarized amount of deals on the Y axis.

Reviewer: ```suggestion
The above code block loads the `ggplot2` package and generates a barchart with the names of the sales reps on the X axis and the summarized amount of deals on the Y axis.
```

ui <- function(id) {
  ns <- NS(id)
  fluentPage(      
    Text("Sales Data Dashboard", variant = "xxLarge"),
    Text("A randomly generated dataset of imaginary sales deals", variant = "large"),
    toggle$ui(ns("toggle")),
    Stack(
      horizontal = TRUE,
      tokens = list(padding = 20, childrenGap = 10),
      span(class = "ms-depth-8", datatable$ui(ns("datatable"))),
      span(class = "ms-depth-8", plot$ui(ns("plot")))
    )
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    data <-toggle$server("toggle")
    datatable$server("datatable", data)
    plot$server("plot", data)
  })
}
```
The code now uses all the other modules that we just created/modified, creates the cards in the UI using `Stack` in a `fluentPage` from the `shiny.fluent` package, calls the modules accordingly and passes on the relevant information to each other in the server to have a working web application.

Once ready, on running the application, we should see the plotly bar chart rendering successfully and also responding to our interactions with the toggle input button.

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_gif_3.gif")

> ```

Reviewer: Please add an empty line below.

```
# in R console
rhino::build_sass()
```

So, this is how the UI should look now -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_gif_2.gif")
```
Note that we now we have space beside the table for additional UI component(s).

## Render a Bar Chart with logic in a helper function

We will now add a bar chart that plots the deal amounts closed by the sales reps. For this, we will code a helper function and keep it inside the `app/logic` folder.

### Helper function in the logic folder of the `rhino` project

* Go to the `app/logic` folder and create a new `.R` file. Lets name it `barchart.R`.
* We shall use the `gglplot2` and `plotly` packages (which we installed in the beginning) for this purpose. 

Here is the code -

```
# app/logic/barchart.R

box::use(
  ggplot2[aes, element_text, geom_bar, ggplot, labs, theme, theme_light],
)

#' @export
generate_barchart <- function(filtered_data) {
  ggplot(filtered_data, aes(x = rep_name, y = deal_amount, fill = rep_name)) +
    geom_bar(stat = "identity") +
    labs(x = "Sales Rep Name", y = "Deal Amount") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 60, hjust = 1))
}
```

> The above code block loads the `ggplot2` package and generates a barchart with names of the sales reps on the X axis and the number of deals on the Y axis.

Reviewer: This is incorrect, Y axis is not the number of deals but the summarized amount of deals.

    filtered_deals <- reactive({
      shiny.fluent::fluentSalesDeals |> filter(
        is_closed | input$includeOpen
      )
    })
    output$datatable <- renderUI({
      DetailsList(items = filtered_deals(), columns = columns)
    })
    output$barchart <- renderPlotly({
      generate_barchart(filtered_deals())
    })
  })
}
```

Here, we have called the `plotly` library, and also the function that we defined in the `app/logic` folder. `plotlyOutput` will help us to render the graph.

Similar to previous usecase, we have added `plotlyOutput` inside the `Stack` so that it is rendered horizontally beside the table as a card. It has the same styling as of the data table's card, hence no need to define the style separately. 

In the server section, we use `renderPlotly` to ask the backend of the application for rendering the barchart.

Once ready, on running the application, we should see the plotly bar chart rendering successfully and also responding to our interactions with the toggle input button.

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_gif_3.gif")
```

## Create rhino modules for toggle button and barchart

> Now that we have successfully developed a working application with a data table, a toggle input and a barchart, let us further create different rhino modules for these features!

Reviewer: ```suggestion
Now that we have successfully developed a working application with a data table, a toggle input, and a barchart, let us further create different Rhino modules for these features!
```

We already have the datatable module with us. We will thus create two more modules - `app/view/toggle.R` and `app/view/plot.R` along with some modifications to `app/view/datatable`. 

> These modules will br called from the main file.

Reviewer: ```suggestion
These modules will be called from the main file.
```

### The `app/view/toggle.R` module -

```
# app/view/toggle.R

box::use(
  dplyr[filter],
  shiny[div, moduleServer, NS, observeEvent, reactive],
  shiny.fluent[Text, Toggle.shinyInput],
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  div(
    Text("A randomly generated dataset of imaginary sales deals", variant = "large"),
    div(
      class = "ms-depth-4",
      Toggle.shinyInput(ns("includeOpen"), label = "Include open deals")
    )
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    filtered_deals <- reactive({
      shiny.fluent::fluentSalesDeals |> filter(
        is_closed | input$includeOpen
      )
    })
    observeEvent(input$includeOpen, {
      filtered_deals()
    })

>     return(filtered_deals())      

Reviewer: This is not correct. `filtered_deals` reactive includes `input$includeOpen` inside, thus it will be recalculated every time it is changed. `observeEvent` is redundant. Also, the module should return `filtered_deals`, not `filtered_deals()` - you want to return a reactive object, not a value.

  }) 
}

```
This module displays a text in the UI along with a toggle input item which the user can interact with. The styling remains the same as we discussed in the previous sections.

It sets up an observer in the server that watches for changes to the `includeOpen` input from the toggle. Whenever this input changes, the `filtered_deals()` reactive expression is called to update the filtered data frame and ultimately the filtered data is returned to the main module.

### The modified `app/view/datatable.R` module

As the `toggle` module returns the filtered data, it is used in the `datatable` module so that the data is displayed on a table accordingly. Hence, our `datatable` module is modified as shown below -

```
# app/view/datatable.R

box::use(

>   shiny[moduleServer, NS, reactive, renderUI, span, uiOutput],

Reviewer: Most of those functions are no longer used here.

#' @export
ui <- function(id) {
  ns <- NS(id)
  plotlyOutput(ns("barchart"))
}

#' @export
server <- function(id, data) {
  moduleServer(id, function(input, output, session) {
    output$barchart <- renderPlotly({
      generate_barchart(data())
    })
  })
}
```
### The modified `app/main.R` module

So far we have created the modules that we need for the app's contents. 

To make them work, we would need to call them in the main module. Here is how the main module has changed -

```
# app/view/main.R

box::use(
  shiny[moduleServer, NS, reactive, span],
  shiny.fluent[fluentPage, Stack, Text],
)

box::use(
  app/view/datatable,
  app/view/toggle,
  app/view/plot,
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  fluentPage(      
    Text("Sales Data Dashboard", variant = "xxLarge"),
    toggle$ui(ns("toggle")),
    Stack(
      horizontal = TRUE,
      tokens = list(padding = 20, childrenGap = 10),
      span(class = "ms-depth-8", datatable$ui(ns("datatable"))),
      span(class = "ms-depth-8", plot$ui(ns("plot")))
    )
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    data <- reactive({
      toggle$server("toggle")      

>     })

Reviewer: Wrapping a module call in a `reactive` is redundant. The module itself should return a reactive value.

This is important because when the UI contains multiple instances of the same element (e.g. multiple tables), each instance must have a unique namespace to prevent naming conflicts in the server function.

In the **`server`** function, the **`ns`** function is not used directly, but its generated namespace is used as an argument for the **`datatable$server()`** function call.
This allows the server to reference the specific instance of the **`datatable`** element that is associated with the current session.

Now run the app.
You should be seeing something like this -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_2.png")
```

### Some improvements

#### Show informative columns with meaningful names

-   We will no more display all the columns that come with the dataset.
    Instead, we will display the `"rep_name", "date", "deal_amount", "city", "is_closed"` columns.

-   But for that we need to make the names more readable.
    So, we will assign proper names for every fields using the `tibble` package -

    ```         
    columns <- tibble(

>         fieldName = c("rep_name", "date", "deal_amount", "city", "is_closed"),

Author: Thanks, I have fixed the indentation issues.

        name = c("Sales Rep", "Date", "Deal Amount", "City", "Is Closed?")

>         )

Author: Sure. Changes applied. Thanks.

```         
install.packages("rhino")
```

Now, initialize a new Rhino application.
You can do it either by using RStudio Wizard or a function `rhino::init()`.

More details on how to create a Rhino app can be found in the [Rhino tutorial](https://appsilon.github.io/rhino/articles/tutorial/create-your-first-rhino-app.html#create-an-initial-application).

You will notice that the working directory now has a proper folder structure along with some other files.
Among those, you will also notice that `renv` is being used to manage the packages/dependencies for the project.
So, let us now install `shiny.fluent` into the project.

```         
renv::install("shiny.fleunt")
```

Using renv for installation benefits us.
In particular, this means that the local cache of package installations is used when possible.
This helps to avoid re-downloading packages that have already been downloaded before, and re-compiling packages from source when a binary copy of that package is already available.

Now, head over to the `dependencies.R` file and add `shiny.fluent` to it as shown below.

```         
library(rhino)
library(shiny.fluent)

> ```

Author: fixed in latest commit

Upon saving the modified file, please take a snapshot of the dependencies so that the `renv.lock` file is updated.

```         
renv::snaphot()
```

Maintain this process whenever we need to install a new package in the project henceforth.

> Note: It is always a good practice to install all required packages at the beginning of your project.

Author: fixed in latest commit

However, for the sake of this tutorial, we introduce you to a new package when the need arises.
For a routine usecase and/or practice, please install all required packages at the beginning.

### Our first development on the app

> -   Open the `app/main.R` file.

Author: removed huge bullet points

>     By default, it has the following code -

Author: excluded in latest commit

>     ```         

Author: fixed in latest commit

### Creating the module

-   In the `app/view/` folder, let us create a new file and name it `datatable.R`

-   This `app/view/datatable.R` file will be used to display the data table on the UI and we will call this module in the `app/main.R` file once done.
    Here is the code

    ```         
    #app/view/datatable.R

    box::use(
      shiny[moduleServer, NS],
      shiny.fluent[fluentPage, Text, DetailsList]
    )

    #' @export
    ui <- function(id) {
      ns <- NS(id)
        fluentPage(
            Text("A randomly generated dataset of imaginary sales deals"),
            DetailsList(items = shiny.fluent::fluentSalesDeals)
        )
    }

    #' @export
    server <- function(id) {
      moduleServer(id, function(input, output, session) {

      })
    }
    ```

`fluentSalesDeals` is a dataset available with the `shiny.fluent` package and to use it, we use `::` for directly fetching the dataset from the library.

So, here we are using [`DetailsList`](https://appsilon.github.io/shiny.fluent/reference/DetailsList.html) function to render the table.

### Calling the module

-   Go to `app/main.R` file and edit it with the following code

```         

> #app/main.R

Author: fixed in latest commit

box::use(
  shiny[moduleServer, NS],
  shiny.fluent[fluentPage, Text]
)

box::use(
  app/view/datatable,
)

#' @export
ui <- function(id) {
  ns <- NS(id)
  fluentPage(
    Text("Hello!", variant = "mega"),
    datatable$ui(ns("datatable"))
  )
}

#' @export
server <- function(id) {
  moduleServer(id, function(input, output, session) {
    datatable$server("datatable")
  }
  )
}
```

Notice here that we have called the `datatable` module by importing it to our main application file first.

> It is now, that the main module will be able to use exported functions from `app/view/datatable.R`.

Author: `datatable` module creation has been explained in the previous section. In this section we are just calling it in main file.

This allows the server to reference the specific instance of the **`datatable`** element that is associated with the current session.

Now run the app.
You should be seeing something like this -

```{r, echo=FALSE}
knitr::include_graphics("images/rhino_fluent_2.png")
```

### Some improvements

#### Show informative columns with meaningful names

-   We will no more display all the columns that come with the dataset.
    Instead, we will display the `"rep_name", "date", "deal_amount", "city", "is_closed"` columns.

-   But for that we need to make the names more readable.
    So, we will assign proper names for every fields using the `tibble` package -

    ```         
    columns <- tibble(
        fieldName = c("rep_name", "date", "deal_amount", "city", "is_closed"),
        name = c("Sales Rep", "Date", "Deal Amount", "City", "Is Closed?")
        )
    ```

-   Also, we need to mention in the `DetailsList()` function inside UI to only display the names we just assigned in the `columns` object.
    Here is how the `ui` function now gets modified -

    ``` r

>     #app/view/databale.R

Author: fixed in latest commit
